File: sorting/insertion_sort.c
// insertion_sort
// https://github.com/ddamddi

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

void insertion_sort(int arr[], int N);
	
int main(int argc, char* argv[]){
	
	int i, N, Nlen;
	int K = 0, counter = 0;
	double Runtime;
	int * arr;
	int * arr_tmp;		// array for memcpy
	char str[30];		// string for fgets()
	FILE * fp;
	clock_t start_time;
	
	// check command line arguments.
	// Ex) <your_program> <N> <input_file> : argc = 3
	if(argc != 3){										
		puts("incorrect command line arguments.");		
		return -1;										
	}
	
	// Check N is integer or not, checking every single character in argv[1]
	// ex) aaa, 12a, a492 etc.
	Nlen = strlen(argv[1]);								
	for(i = 0; i< Nlen; i++){							
		if(argv[1][i] < '0' || argv[1][i] >'9'){		
			puts("N is not integer.");
			return -1;
		}
	}
	
	N = atoi(argv[1]);
	
	fp = fopen(argv[2],"rt");
	
	// check file open error.
	if(fp == NULL){			
		puts("file open fail.");
		return -1;
	}
	
	
	while(!feof(fp)){
		// count K (the number of integer in file)
		fgets(str, sizeof(str), fp);
		K++;
	}
	
	// N > K, sort K integers
	// N <= K, sort N integers.
	if(N > K)
		N = K;
	
	arr = (int*)malloc(sizeof(int)*N);
	fseek(fp,0,SEEK_SET);
	
	for(i = 0 ; i < N; i++){							
		fgets(str, sizeof(str), fp);
		arr[i] = atoi(str);
	}
	
	fclose(fp);					
	arr_tmp = (int*)malloc(sizeof(int)*N);
	
	// timer start
	start_time = clock();							
	do {
		// memory copy to record precise running time
		memcpy(arr_tmp, arr, sizeof(int)*N);
		counter++;
		insertion_sort(arr_tmp, N);
	} while(clock() - start_time < 1000);
	
	// timer end & calculate time.
	Runtime = (double)(clock() - start_time) / (CLOCKS_PER_SEC/1000);
	Runtime = (double)Runtime / counter;
	
	// Print all array
	for(i=0 ; i < N ; i++){
		printf("%d\n", arr_tmp[i]);
	}

	printf("Running time : %lf ms\n", Runtime);
 	
	free(arr);
 	free(arr_tmp);
	return 0;
}

void insertion_sort(int arr[], int N){
	
	int i, j, temp;
	
  	for(i = 1; i < N; i++){
  		
  		j = i;
  		temp = arr[j];
  		
  		// compare arr[i] with neighboring arr[i-1]
  		// if exchange values, go left and continue to compare
  		while(--j >= 0 && temp > arr[j]){
  			
			arr[j+1] = arr[j];
			arr[j] = temp;
		}
  	}
}

File: insertion_sort.java
// Student ID : 20144320
// Student Name : Kim Gyeong-Hyeon

import java.io.BufferedReader;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;
import java.util.Scanner;

public class insertion_sort {

	public static void insertionSort(int[] list){
		
		int i, j, temp;
		
	  	for(i = 1; i < list.length; i++){
	  		
	  		j = i;
	  		temp = list[j];
	  		
	  		// compare arr[i] with neighboring arr[i-1]
	  		// if exchange values, go left and continue to compare
	  		while(--j >= 0 && temp > list[j]){
	  			
				list[j+1] = list[j];
				list[j] = temp;
			}
	  	}
	}

	
	public static void main(String[] args) {
		
		// check incorrect command line arguments
		if(args.length != 2) {
			System.out.println("incorrect command line arguments.");
			System.exit(-1);
		}
		
		int N = 0;
	
		try {
			N = Integer.parseInt(args[0]);
		} catch (Exception e) {
			// Check whether N is integer or not
			System.out.println("N is not integer.");
			System.exit(-1);
		}
		
		
		int count = 0;
		
		// args[1] is Filename
		File file = new File(args[1]);
		
		// count K (the number of integer in file)
		try {
			BufferedReader bufReader = new BufferedReader(new FileReader(file));
			String line = "";
			while((line = bufReader.readLine()) != null)
				count++;
			bufReader.close();
		} catch (Exception e) {
			System.out.println("File open error.");
		}
		
		// N > K, sort K integers.
		// N <= K, sort N integers.
		if(N > count)
			N = count;
		
		int[] arr, arr_temp;
		arr = new int[N];
		arr_temp = new int[N];
		
		// Save K number integer from file to arr
		try {
			Scanner scanner = new Scanner(file);
			for(int i = 0; i < N; i++) {
				arr[i] = Integer.parseInt(scanner.nextLine());
			}
		} catch (Exception e) {
			System.out.println("File open error.");
		}
		
		// counter for running time
		int counter = 0;
		long startTime = System.currentTimeMillis();
		
		do {
			System.arraycopy(arr, 0, arr_temp, 0, N);
			insertionSort(arr_temp);
			counter++;
		} while( (System.currentTimeMillis() - startTime) < 1000 );
		// Running time < 1000ms, Run several times for precise value.
		
		float runningTime = (float)(System.currentTimeMillis() - startTime) / counter;
	
		// print sorted array
		for(int i = 0; i < N; i++) {
			System.out.println(arr_temp[i]);
		}
		// running time
		System.out.println("running time : " + runningTime + " ms");
		
	}
}
